src/workerd/api/sockets.h

```diff
@@ -69,27 +72,48 @@ struct SocketOptions {
 struct InitData {
   jsg::Ref<ReadableStream> readable;
   jsg::Ref<WritableStream> writable;
-  IoOwn<kj::PromiseFulfillerPair<void>> closeFulfiller;
+  IoOwn<kj::PromiseFulfillerPair<bool>> closeFulfiller;
+  IoOwn<kj::PromiseFulfillerPair<void>> jsCloseFulfiller;
+  kj::ForkedPromise<void> jsCloseFulfillerFork;
 };
 
 class Socket: public jsg::Object {
 public:
-  Socket(InitData data)
+  Socket(jsg::Lock& js, InitData data)
       : readable(kj::mv(data.readable)), writable(kj::mv(data.writable)),
-        closeFulfiller(kj::mv(data.closeFulfiller)) {};
-  Socket(kj::Promise<kj::Own<kj::AsyncIoStream>> connectionPromise);
+        closeFulfiller(kj::mv(data.closeFulfiller)),
+        jsCloseFulfiller(kj::mv(data.jsCloseFulfiller)),
+        jsCloseFulfillerFork(kj::mv(data.jsCloseFulfillerFork)) {
+    auto& context = IoContext::current();
```

Comment:
This constructor body is too complex to actually be inlined. As a rule, only inline-able code and template should appear in headers; everything else should go in the `.c++` file. So this constructor by should be moved to `sockets.c++`.